### ISSUE TITLE
Image Preview Generation Fails

### ISSUE THREAD

**Issue by shasderias** [nextcloud/nextcloud-snap#24]
I am encountering the following error with installations of Nextcloud. The error occurs on both revision 17  (sudo snap install nextcloud) and revision 22 (the revision with https support installed with sudo snap install nextcloud --beta).

After creating an admin account, navigating to the Photos folder using the File manager and attempting to open a photo returns the following error.

"Error! Could not generate a preview of this file."

The error appears to be caused by a call to the following URL (which returns error code 500):
index.php/apps/gallery/preview/6?width=2000&height=2000…eWQNFJQgqfDUKLAUuVmk%3D%3AsrRCRYS4hqevSFghRu6V%2BUcX3RlCuidXL2oX9bw6EO0%3D 

No error is logged at the default log level. Configuring Nextcloud to log everything yields the following log entries:

Debug	core	OC_Image->fixOrientation() Orientation: -1	2016-08-19T14:15:27+00:00	admin
Debug	core	OC_Image->fixOrientation() Image is not a JPEG.	2016-08-19T14:15:27+00:00	admin
Debug	core	OC_Image->loadFromFile, JPG images not supported: /var/snap/nextcloud/common/nextcloud/data/admin/files/Photos/San Francisco.jpg	2016-08-19T14:15:27+00:00	admin

Uneducated guess would be that php is not compiled with libjpeg.

**Comment by kyrofa** [comment]
> Uneducated guess would be that php is not compiled with libjpeg.

That sounds like a very educated guess! Sounds like we're missing a flag [here](https://github.com/nextcloud/nextcloud-snap/blob/master/snapcraft.yaml#L87). Initial research leads me to believe it's `--with-jpeg-dir=<wherever libjpeg gets installed>` with perhaps `libjpeg-dev` as a `build-package`. Though I'm not sure whether we need jpeg8 or jpeg9, any idea?

**Comment by oparoz** [comment]
```
        configflags:
...
          - --with-gd
          - --enable-gd-native-ttf
          - --with-jpeg-dir=/usr/lib
          - --with-freetype-dir=/usr/lib
    stage-packages:
...
      - libfontconfig1
      - libfreetype6
      - libjpeg8
    build-packages:
...
      - libgd-dev
```

Not sure if jpeg9 is a better option but it was working fine with jpeg8

**Comment by kyrofa** [comment]
@oparoz are `--enable-gd-native-ttf` and `--with-freetype-dir` required for this?

**Comment by oparoz** [comment]
Not for images, but it's just the proper way to configure GD for a hosting environment so that app developers are less limited. Basic image and text manipulation in PHP is usually a given.

**Comment by kyrofa** [comment]
> Not for images, but it's just the proper way to configure GD for a hosting environment so that app developers are less limited. Basic image and text manipulation in PHP is usually a given.

libjpeg has an obvious use-case (this bug). Do we have a use-case for those other capabilities in Nextcloud? I don't see a big advantage to enabling things "just because"-- this all takes space and increases potential attack surfaces.

**Comment by oparoz** [comment]
The most obvious one is watermarking. If someone writes an app which adds a copyright on the image, PHP needs to have access to special methods and to system fonts. Hmmm... This last part rings a bell when trying to configure Imagemagick... fonts might not be easily accessible unless we set an env variable and we can't easily test this functionality since the root path is read-only.
Let's move that to another PR then.

**Comment by kyrofa** [comment]
@shasderias @oparoz mind giving #25 a look/test? Details for how to test are over there.

**Comment by kyrofa** [comment]
This was resolved in #25. Thank you for the excellent bug report!
